### PR TITLE
build: speedup tsc

### PR DESCRIPTION
**Pull request by trim21**
This is a (almost) type only change, it change the type of `TriggerSet` only. Our core file are using `LooseTriggerSet` and `LooseTrigger` so they are not affected.

This patch make `trigger.type` required for all cactbot built-in net triggers and merge `BaseTrigger` and `PartialNetRegexTrigger` to help tsc narrow trigger type.

about why merging types is explained here: https://github.com/quisquous/cactbot/issues/4508#issuecomment-1298080356

## result

`npx tsc` 35s -> 20s
`npx eslint .` 1m6.862s -> 40.713s
`npm test` 38s -> 20s
`npm run start` no improvement.

---

I know the team of swc is trying to port tsc to rust, wish them good luck…

### CODE REVIEW

File: ui/raidboss/popup-text.ts [hunk 1]
```diff
@@ -762,7 +762,7 @@ export class PopupText {
               trigger.localNetRegex = Regexes.parse(localeNetRegex);
               orderedTriggers.push(trigger);
               found = true;
-            } else if (defaultNetRegex) {
+            } else if (defaultNetRegex !== undefined) {
```
Comment (trim21) [inline ui/raidboss/popup-text.ts:765]: this trigger eslint error about "always true"

**Comment by trim21** [comment]
trace after this patch:

```text
~/proj/cactbot (speedup-tsc-a-little) # time npm run tsc-no-emit -- --generateTrace ../cactbot-tracing && npx analyze-trace ../cactbot-tracing --force-millis 200

> cactbot@0.27.21 tsc-no-emit
> tsc --noEmit --generateTrace ../cactbot-tracing


real    0m20.240s
user    0m0.046s
sys     0m0.166s
Hot Spots
├─ Check file c:\users\trim21\proj\cactbot\node_modules\typescript\lib\lib.dom.d.ts (366ms)
└─ Check file c:\users\trim21\proj\cactbot\ui\raidboss\data\05-shb\eureka\delubrum_reginae_savage.ts (208ms)
   └─ Check variable declaration from (line 94, char 7) to (line 3712, char 2) (185ms)
      └─ Check expression from (line 94, char 38) to (line 3712, char 2) (169ms)
         └─ Check expression from (line 2601, char 20) to (line 3711, char 4) (126ms)

No duplicate packages found
```

` Check expression from (line 2601, char 20) to (line 3711, char 4) (126ms)` is a huge `timelineReplace`

File: ui/raidboss/popup-text.ts (hunk 1 above)
Comment (trim21) [reply ui/raidboss/popup-text.ts:765]: we may need to refactor the code for processing triggers, it would be much easy to maintain if it's pure function.

Comment (trim21) [reply ui/raidboss/popup-text.ts:765]: And I find the reason why it's not working:

```ts
export type LooseTriggerSet =
  & Exclude<Partial<TriggerSet<RaidbossData>>, 'triggers' | 'timelineTriggers'>
  & {
    /** @deprecated Use zoneId instead */
    zoneRegex?:
      | RegExp
      | { [lang in Lang]?: RegExp };
    triggers?: LooseTrigger[];
    timelineTriggers?: LooseTimelineTrigger[];
  };
```

here it's not `Exclude`, it should be `Omit`

File: ui/raidboss/popup-text.ts [hunk 2]
```diff
@@ -718,7 +718,7 @@ export class PopupText {
           // time later.  This will clobber each time we
           // load this, but that's ok.
           trigger.filename = setFilename;
-          const id = trigger.id;
+          const id = trigger.id ?? trigger.toString();
```
Comment (trim21) [inline ui/raidboss/popup-text.ts:721]: here `id` is optional, and it's option in user trigger, I don't know how to handle there...

Comment (trim21) [reply ui/raidboss/popup-text.ts:721]: And I found it's a much more complicated type issue here https://github.com/quisquous/cactbot/issues/5039

fixing the `LooseTriggerSet` cause many lint error.

**Comment by trim21** [comment]
This PR is ready to merge wither https://github.com/quisquous/cactbot/issues/5039 is fixed or not

**Comment by quisquous** [comment]
Thanks for the investigation and the fix.  That's really smart to combine the types like that to help out tsc.

**Pull request merged**